src/Microsoft.VisualStudio.LanguageServices.Razor/Editor/RazorTextViewConnectionListener.cs

```diff
@@ -0,0 +1,156 @@
+            TextBufferFactoryService textBufferFactoryService,
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace)
+        {
+            if (textBufferFactoryService == null)
+            {
+                throw new ArgumentNullException(nameof(textBufferFactoryService));
+            }
+
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+
+            _textBufferFactoryService = textBufferFactoryService;
+            _workspace = workspace;
+
+            _foregroundDispatcher = workspace.Services.GetRequiredService<ForegroundDispatcher>();
+        }
+
+        // This is only for testing. We want to avoid using the actual Roslyn GetService methods in unit tests.
+        internal RazorTextViewConnectionListener(
+            ForegroundDispatcher foregroundDispatcher,
+            TextBufferFactoryService textBufferFactoryService,
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace)
+        {
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (textBufferFactoryService == null)
+            {
+                throw new ArgumentNullException(nameof(textBufferFactoryService));
+            }
+
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _textBufferFactoryService = textBufferFactoryService;
+            _workspace = workspace;
+        }
+
+        public Workspace Workspace => _workspace;
+
+        public void SubjectBuffersConnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (textView == null)
+            {
+                throw new ArgumentException(nameof(textView));
+            }
+
+            if (subjectBuffers == null)
+            {
+                throw new ArgumentNullException(nameof(subjectBuffers));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            for (var i = 0; i < subjectBuffers.Count; i++)
+            {
+                var textBuffer = subjectBuffers[i];
+                if (!textBuffer.IsRazorBuffer())
+                {
+                    continue;
+                }
+
+                if (!_textBufferFactoryService.TryGetDocumentTracker(textBuffer, out var documentTracker) ||
+                    !(documentTracker is DefaultVisualStudioDocumentTracker tracker))
+                {
+                    Debug.Fail("Tracker should always be available given our expectations of the VS workflow.");
+                    return;
+                }
+
+                // REVIEWERS: Thinking of removing tracker.TextViewsInternal and replacing it with two methods:
```

Comment:
👀

Reply:
Add those methods on `DefaultVisualStudioDocumentTracker` - not on the abstract class and we're fine 👍